### DISCUSSION TITLE
DHT22 bouncing Humidity parameter

### DISCUSSION THREAD

**Discussion by cleverpal-ai** [arendst/Tasmota#16134] [Q&A] [answered]
I'm using a bunch of Sonoff Mini's with DHT22 (AM2301) with Tasmota Firmware 12.0.2.
I need to track a Humidity value and proceed with further actions. But Humidity value changes too fast and in vast range (75-99-81- and so on, can change within 10 seconds). The real humidity doesn't change that fast. I have a reference unit of Shelly 1PM + Add-on (with same sensors), they report a stable Humidity parameter. Temperature value, though, doesn't jump as Humidity.
I use 4.7K pull-up resistor. Tried to use 4.7+4.7 couple, no changes on that figure.
Is there anything I need to change in Firmware to get it work properly?

P.S. Changed 4 sensors (all new), each behaves same way with Tasmota.

**Comment by barbudor** [comment]
Why openning various discussions about your failures with DHT/AM23xx sensors
The answer is all the same : the proper location for those DHT/AM23xx sensors are in the trash
Move to BME280 and use a NodeMCU or Mini-D1, not a Sonoff which have barely enough PSU to manage it's basic job

**Reply by cleverpal-ai** [reply]
@barbudor I did not find any relevant topics like this.
Anyway, using the opportunity, will ask you - why do the same sensors (AM2301, various options, see below) show such a high humidity on Sonoff flashed by Tasmota? I mean - in standard Sonoff (this time TH16) firmware in the same place where you see 90+% it shows 60% (which is true). Can this figure be calibrated in Tasmota? 

OK, I will get BME280, but I need to use Sonoff due to compact mains power terminal block and 220V output in the same place. D1 Mini would need separate module that takes impermissibly large space for my installation.
Do you think it's possible to connect 2 sets BME280 to 1 unit Sonoff Mini?

Thanks for your reply & help!

**Reply by cleverpal-ai** [reply]
Besides, I'm using a bunch of AM2301's in my installations with Shelly, and they for years like a charm. Visually they are much more attractive compared to a bare BME280 :)

**Reply by cleverpal-ai** [reply]
![image](https://user-images.githubusercontent.com/62514634/182448812-d5f8ed56-240c-4fa5-8a68-eec3314a4a2a.png)
![photo_2022-08-02_21-37-36 (2)](https://user-images.githubusercontent.com/62514634/182449163-a97fec8d-804f-472b-b579-a681a8934436.jpg)

**Reply by cleverpal-ai** [reply]
![photo_2022-08-02_21-37-36](https://user-images.githubusercontent.com/62514634/182449271-a661dd48-12f2-4d19-a023-1b80e6cd4d6f.jpg)

**Comment by barbudor** [comment, marked as answer]
BME280 uses I2C so 2 wires
Tasmota's implementation supports up to 2 units with addresses 0x76 and 0x77: don't forget you need to configure your BME module to set a different address. Generaly there are some pads to solder.

**Reply by cleverpal-ai** [reply]
@barbudor So I will need to use SDA & SDL and attach each of them to a separate GPIO of Sonoff Mini, right?
As far as I understand the rest of the contacts (CSB, SDO) will remain free.

Could you please elaborate how to configure such sensor with Tasmota?
Thanks!

![image](https://user-images.githubusercontent.com/62514634/182540822-5a267a54-2166-4831-8640-350fe875f94e.png)

**Reply by barbudor** [reply]
Yes, you need VCC, GND, SCL, SDA
BME280 driver is included by default in tasmota-sensor (ESP8266) or tasmota (ESP32)
You need to check that the proper I2C driver is enabled using command `i2cdriver`
That will list you a lot of numbers which matchs the drivers' index at [I2CDEVICES](https://tasmota.github.io/docs/I2CDEVICES/) 
You need driver 10 to be enabled (no `!` sign in front of it) and all others to be disabled to avoid mis-detection
This can be done using a backlog of multiple `i2cdriver` commands like below:
```
08:07:40.604 CMD: i2cdriver
08:07:40.610 RSL: RESULT = {"I2CDriver":"7,8,9,10,11,12,13,14,15,17,18,20,24,29,31,36,41,42,44,46,48"}
```
```
backlog i2cdriver7 0;i2cdriver8 0;i2cdriver9 0;i2cdriver11 0;i2cdriver12 0;i2cdriver13 0;i2cdriver14 0;i2cdriver15 0;i2cdriver17 0;i2cdriver18 0;i2cdriver20 0;i2cdriver24 0;i2cdriver29 0;i2cdriver31 0;i2cdriver36 0;i2cdriver41 0;i2cdriver42 0;i2cdriver44 0;i2cdriver46 0;i2cdriver48 0 
```
After reboot:
```
08:10:43.458 CMD: i2cdriver
08:10:43.465 RSL: RESULT = {"I2CDriver":"!7,!8,!9,10,!11,!12,!13,!14,!15,!17,!18,!20,!24,!29,!31,!36,!41,!42,!44,!46,!48"}
```
Disabling un-used drivers avoid mis-detection of a sensor from another

By default the sensor will be at address 0x76 (automatically detected by Tasmota driver)
To use a 2nd sensor in parrallel, you need to connect the SDO pin to VCC in order to set the 2nd sensor to address 0x77
When 2 sensors are detected, Tasmota will always assign BME280-1 to the one at 0x76 and BME280-2 to the one at 0x77
If only one is detected, it will be BME280 wether it's address is 0x76 or 0x77

Hope this help

**Reply by barbudor** [reply]
Just in case you want more than 2 : it is not possible to have more than 2 BME280 as it supports only 2 addresses (0x76 and 0x77)
However a nice alternative is HDC1080 which use address 0x40 so that could add a 3rd Temp/Hum sensor without conflict. Except that HDC1080 drivers (index 45) is not included in default buils of Tasmota so you will need to self-compile.

**Reply by cleverpal-ai** [reply]
I bought this type of sensors
![image](https://user-images.githubusercontent.com/62514634/183048921-747991f9-c672-4b08-b198-c36baf61eddf.png)

They don't have an SDO pin. Do I need to make any modifications to above in order for the 2nd sensor to assign a different address? Or just attach SDA/SCL to a free GPIOs?

**Comment by cleverpal-ai** [comment]
@barbudor 
I've compiled "tasmota-lite" with I2C driver enabled, but after I disabled I2CDriver7 and attempted to enable I2CDriver10 I receive the following result:

```
09:27:05.294 CMD: I2CDriver10 1
09:27:05.300 MQT: stat/CleverDryerSonoff_653743/RESULT = {"I2CDriver":"!7"}
```

Cannot get I2CDriver10 to work.
Do I need to compile something special to make it work?

**Reply by Jason2866** [reply]
Yes, do not use tasmota-lite as base for self compiled customized versions.
Free configurable are only the variants `Tasmota` and `Tasmota32`.

**Reply by Jason2866** [reply]
Btw. DHT22 does work with Tasmota in general. Since there are many bad clones sold the result is as expected when cheap stuff (in every aspect) is used. If you want to use buy tons from different sellers and you will probably get some which are working somehow. When real reproducable values are needed avoiding DHT is the way...

**Reply by cleverpal-ai** [reply]
@Jason2866 
My device (Sonoff Mini) has some limitations in terms of memory, thus I initially preferred `tasmota-lite`.
So you suggest to use a standard `Tasmota` for BME280 to work, right? I can try to compile and .tar.gz'ip it, maybe it will work out.

I used a bunch of ASAIR AM2301 (DHT22) sensors with Shelly relays, all of them show correct humidity parameters. In Tasmota (3 different controllers like ESP32, Sonoff Mini, TH16) the same sensor shows delirious figures.
Of course, I will try BME280, but I2C has cable length limitations (my case - about 1.5m length from controller to a sensor), thus I will need to additionally slow down the bus for it. Ultimately I need to get realistic humidity figures from 2 sensors, connected to 1 controller.

**Reply by cleverpal-ai** [reply]
`#define USE_BMP` selected in source for `tamota-lite` solved the issue. 
Now will try using 2nd sensor

**Reply by cleverpal-ai** [reply]
Just FYI, here's 2 sensors within 3 cm distance, have a look at the discrepancy of figures...

![image](https://user-images.githubusercontent.com/62514634/183044914-ffcd0302-52a7-4c64-899e-64d24fe42916.png)

**Reply by sfromis** [reply]
With ESP8266 standard of at least 1MB of flash, there is no "limitations in terms of memory" with ESP8266, and all the precompiled Tasmota binaries can work, as long as you make sure to use the compressed .gz binaries when upgrading. Memory-wise you get no functional advantage starting from tasmota-lite.

**Reply by Jason2866** [reply]
The difference is normal 1°C. For humidity there are many aspects. Have both sensors a burn in time >1 week? Do the have been stored well before using? In general i doubt always humidity from DHT...
Get some more BMP/BME or other actual I2C Temp/Humi sensors and compare this ones.
To change the address of your sensor. Cut the bridge beetween the solder pads (left middle) and connect the middle and the right solder pad.

**Reply by cleverpal-ai** [reply]
@Jason2866 Sorry, I didn't quite get what bridge I should cut in this type of sensor?
![image](https://user-images.githubusercontent.com/62514634/183064386-c8a19331-e86e-4ed4-9fa2-1a2cd5c46b91.png)
And what pins should I interconnect then? Can you elaborate more in details, please?

Thanks!

**Reply by barbudor** [reply]
![IMG_20220805_131324](https://user-images.githubusercontent.com/2996042/183066208-362bb6de-7475-4d50-a8e4-866a3da0d8a1.jpg)

**Reply by cleverpal-ai** [reply]
Thanks! I did it.
![image](https://user-images.githubusercontent.com/62514634/183071500-be174204-d4fb-4e30-a6d9-9f54530df747.png)

But this second "modified" device doesn't seem to be discovered by Tasmota.
What else may be the reason?

P.S. Measuring impedance between left and middle solder pad still shows some interconnection. Is this normal?

**Reply by barbudor** [reply]
Those pins are connected to the BME pins
If you get anything non-zero that should be ok
You can double check by switching the red and black probe of your meter. If you get different values it is ok
One should be quite high resistance
Other way could be smaller (but non zero) because of clamping diodes in the BME
If you get very small identical values in both ways, may be you didn't cut properly

**Reply by cleverpal-ai** [reply]
When I swap the probes, the impedance doesn't change. It's almost 0 ohm (0.3 to be more precise), both ways.
I used the needle file to cut deeper (I'm sure there's no physical contact left), but now I measure between already connected mid+right pads and the left one. When I connect this modified BME separately, it also doesn't work. Maybe there's a way to check what address is assigned? Anything else to do with it?

**Reply by cleverpal-ai** [reply]
Also it's strange, because my TX/RX GPIO(01/03 respectively) do not work under this config with either of 2 BME280's. 
![image](https://user-images.githubusercontent.com/62514634/183083728-3177fa4c-1fbb-4412-8311-cb3aa360724c.png)

![image](https://user-images.githubusercontent.com/62514634/183083688-d1631787-dd73-4183-9b09-c94842d247eb.png)

Now I can use only GPIO(02/00) and only unmodified sensor works while connected there. 
Maybe it's 'tasmota-lite', again? I'll try flashing Tasmota.

**Reply by cleverpal-ai** [reply]
UPD: Flashed full 'Tasmota' - same story.
When using above config, and both sensors (modified + unmodified) connected, nothing shows up.
If I leave one sensor (unmodified) connected to GPIO2/GPIO0 or GPIO1/GPIO3, it works.

**Reply by sfromis** [reply]
You can only have one set of I2C pins, and all I2C devices need to be connected to the same SCL+SDA. That's why you need to strap one BME280 to have a different address on the bus.

**Reply by cleverpal-ai** [reply]
@sfromis 
OK, I connected both BME280 to the same SCL+SDA, but... only one found still
`{"I2CScan":"Device(s) found at 0x76"}`

**Reply by barbudor** [reply]
When connecting only one, do you always get 0x76 in the scan?
The modified one should be reported at 0x77

**Reply by cleverpal-ai** [reply]
Yes, only non-modified gets 0x76, the modified one is not detected at all.

**Reply by barbudor** [reply]
Broken ?

**Reply by cleverpal-ai** [reply]
I don't know. Should this modified unit be detected while connected alone? If Yes, then I guess it might be broken.

**Reply by Jason2866** [reply]
Yes, when connected as only i2c device it should be detected and working.

**Reply by cleverpal-ai** [reply]
I've unsoldered mid+right pads and "modified" device began to work...
But together they don't work...

**Reply by cleverpal-ai** [reply]
Soldered mid+right back and again "modified" BME doesn't work.

**Reply by cleverpal-ai** [reply]
Tried to solder mid+right better and isolate left+mid pads more. Somehow I managed to make "modified" device, but it also works only while alone. And unmodified works alone. As soon as second gets connected, I get figures stuck & null.
![image](https://user-images.githubusercontent.com/62514634/183140292-99061d10-9df0-4465-bc1e-28827c49837a.png)

**Reply by cleverpal-ai** [reply]
In both cases when devices are connected I get:
`{"I2CScan":"Device(s) found at 0x76"}`

Can it be insufficient power or so?
I'm experimenting from USB UART converter and Sonoff is powered from 3.3V.

**Reply by barbudor** [reply]
When both are connected?
So the modification doesn't work and both are conflicting with each other as they are on the same address. Which explain the bad values

**Reply by cleverpal-ai** [reply]
> When both are connected? So the modification doesn't work and both are conflicting with each other as they are on the same address. Which explain the bad values
Yes, right.
I tried to undo the soldering with first and did "the trick" with another sensor - same story.
When I solder mid+right & cut left+mid sensor stops working at all being connected alone to the board.
Something is wrong here, and I have no idea what exactly.
Followed exactly this guide:

![image](https://user-images.githubusercontent.com/62514634/183144523-e303e126-b639-4abb-b816-47753bc5a735.png)

**Reply by sfromis** [reply]
This worked just fine when I did it on that BME280 board.

**Reply by cleverpal-ai** [reply]
Well, I tried with 2nd new BME280. As soon as modified it, it cannot be detected by Tasmota at all.
As far as I understand any modified device should appear with the address 0x77.

**Reply by sfromis** [reply]
Dirty, and messy soldering, but works: 
![image](https://user-images.githubusercontent.com/47082390/183155943-b1c45fd9-2dac-4193-842b-e8a7a9dbb476.png)
IIRC, I moved that module to the dust drawer due to it being a bit out of whack when it comes to values.

**Reply by cleverpal-ai** [reply]
> 

Yes, exactly as I did. 
Somehow my modules refuse to assign 0x77 address.
Here's the second module. Tried to cut very carefully and solder too...
![image](https://user-images.githubusercontent.com/62514634/183157285-401e4d93-1032-43a3-a52f-1bc3690bc676.png)

**Comment by cleverpal-ai** [comment]
It seems to be my issue, guys....:
https://forum.arduino.cc/t/trouble-changing-i2c-address-on-bme280/907865/16

I'm thinking about combining usage of my still operable BME280 together with
[AHT10](https://www.electroschematics.com/temperature-sensor/)

What do you think? Will Tasmota be able to assign 0x76 to my standard BME280 sensor & 0x38 AHT10, while using them on one line (parallel SCL+SDA)? Tired of useless soldering already....

**Reply by sfromis** [reply]
Anything but AHT10; that one is known to be a pig on the I2C bus, not working well with other sensors, specifically including BME280. More generally, with sensors well-behaved on the bus, and unique addresses, Tasmota has no trouble dealing with different sensor drivers at the same time.

**Comment by cleverpal-ai** [comment]
Do you recommend HTU21? 
Will it work together with BME280?

There is not a big selection of such devices around here.

**Reply by sfromis** [reply]
Not tried that one; have heard nothing bad about it, probably ok.
No address conflict with BME, but sometimes it is worth checking
https://github.com/arendst/Tasmota/blob/development/I2CDEVICES.md

I'm mostly using BME280; I like to have atmospheric pressure too, even if it may not be terribly important.

**Comment by barbudor** [comment]
I recommend HDC1080

**Reply by cleverpal-ai** [reply]
Could not find this one in my country. 
By the way, I ordered two more BME280's with SDO pin, maybe this time it will be able to change address.

**Comment by cleverpal-ai** [comment]
Gentelmen, HTU21 + BME280 work together like a charm
![image](https://user-images.githubusercontent.com/62514634/183254438-438c22cd-97ee-4a68-b1c3-58ff4ea74e21.png)
`{"I2CScan":"Device(s) found at 0x40 0x76"}`

**Reply by Jason2866** [reply]
The values from the two sensors are very close.

**Reply by cleverpal-ai** [reply]
> The values from the two sensors are very close.

Yes, Now they're almost identical (sensors are within 10 cm distance) as HTU cooled down from soldering.
I have also AHT10 and tried to activate `I2CDriver43 1` in accordance with Tasmota docs, but I see no driver with index "43"
`{"I2CDriver":"!7,!8,9,10,!11,!12,!13,!14,!15,!17,!18,!20,!24,!29,!31,!36,!41,!42,!44,!46,!48"}`
My firmware - is official Tasmota-Sensors.

**Reply by sfromis** [reply]
If you want to fight with the AHT10, you need a custom build to include support for it, in accordance with Tasmota docs.
https://tasmota.github.io/docs/Firmware-Builds/
Not a good candidate for taking up space in default builds, due to the problematic nature of it.

**Comment by cleverpal-ai** [comment]
Gentlemen, would you please assist me once again?
I'm now using WT32-ETH01 board with Tasmota32.
Connected BME280 and BH1750 using one I2C channel in parallel as before.
I've got drivers 10 and 11 enabled. My BME280 does not show up when I have only driver 10 enabled. If I enable driver 11 (BH1750), I see both sensors data, but BME280 figures reman unchanged all the time (same temperature, humidity and pressure). 
What can be wrong?

**Reply by sfromis** [reply]
My experience is that "power quality issues" may mean that BME280 can lock up. Things like not enough current, or unstable power. Dubious wiring can also make trouble, even if it may look reasonable. Haven't looked into low level details of what happens...

**Reply by cleverpal-ai** [reply]
I used 2 different PSU, incl 5V 2A mobile charger via USB-UART and separate PSU with adjustable voltage and current. Experimented with powering ESP32 with slightly higher than 5V voltage. No changes.
Now I see that BME280 is identified via `i2cscan`, however no data is displayed by EPS at all. At the same time BH1750 works.

**Comment by cleverpal-ai** [comment]
Just in case:
`I2cscan`
`23:12:37.317 RSL: RESULT = {"I2CScan":"Device(s) found at 0x23 0x76"}`

**Reply by cleverpal-ai** [reply]
I am using long Ethernet cable for communication between ESP32 & devices. Maybe this could be the reason of buggy behaviour? Cannot install ESP32 closer to sensors... I read somewhere that by slowing down bus speed, cable length may be increased significantly. Can you advise how to solve?

**Reply by barbudor** [reply]
I2C is meant to be used at the board level, not on long distance: a few 10 cm max
To solve, put an ESP close to each sensor and use wireless connection

The whole story about using cheap ESP device is to go wireless and remove wires.
Stop trying to use parts for something they are not designed for

Or use wireless sesors like Mija (Bluetooth) or Aqara (Zigbee)

**Reply by sfromis** [reply]
If you absolutely want *way too long* I2C connections, you can get I2C extenders, but that is likely to cost more than extra ESP modules. Not really recommended.
